src/components/CheckBoxFilterContainer.ts

```diff
@@ -0,0 +1,203 @@
+        // construct constraint based on checked
+        const constraint = this.getConstraint(isChecked);
+        if (this.dataSourceHelper)
+            this.dataSourceHelper.setConstraint(this.props.friendlyId, constraint);
+    }
+
+    private getConstraint(isChecked: boolean) {
+        const { targetListView } = this.state;
+        if (targetListView && targetListView._datasource) {
+            const attribute = isChecked ? this.props.attribute : this.props.unCheckedAttribute;
+            const filterBy = isChecked ? this.props.filterBy : this.props.unCheckedFilterBy;
+            const constraint = isChecked ? this.props.constraint : this.props.unCheckedConstraint;
+            const attributeValue = isChecked ? this.props.attributeValue : this.props.unCheckedAttributeValue;
+
+            if (filterBy === "XPath") {
+                return constraint.indexOf(`[%CurrentObject%]'`) !== -1
+                    ? constraint.replace(`'[%CurrentObject%]'`, this.props.mxObject.getGuid())
+                    : constraint;
+            } else if (filterBy === "attribute") {
+                return this.getAttributeConstaint(attribute, attributeValue);
+            } else {
+               return "";
+            }
+        }
+    }
+
+    private getAttributeConstaint(attribute: string, attributeValue: string): string | HybridConstraint {
+        const { targetListView } = this.state;
+
+        if (window.device) {
+            const constraints: HybridConstraint = {
+                attribute,
+                operator: "contains",
+                path: this.props.listViewEntity,
+                value: attributeValue
+            };
+            return constraints;
+        }
+        if (targetListView && targetListView._datasource) {
+            const mxObject = targetListView._datasource._pageObjs[0];
+            if (mxObject && mxObject.isEnum(attribute)) {
+                return `[${attribute}='${attributeValue.trim()}']`;
+            } else if (mxObject && mxObject.isBoolean(attribute)) {
+                return `[${attribute} = ${attributeValue.trim().toLowerCase()}()]`;
+            } else {
+                return `[contains(${attribute},'${attributeValue}')]`;
+            }
+        }
+    }
+
+    private connectToListView() {
+        const filterNode = findDOMNode(this).parentNode as HTMLElement;
+        const targetNode = Utils.findTargetNode(filterNode);
+        let targetListView: ListView | null = null;
+
+        if (targetNode) {
+            targetListView = dijitRegistry.byNode(targetNode);
+            if (targetListView) {
+                if (!targetListView.__customWidgetDataSourceHelper) {
+                    try {
+                        targetListView.__customWidgetDataSourceHelper = new DataSourceHelper(targetListView);
+                    } catch (error) {
+                        console.log("failed to instantiate DataSourceHelper \n" + error); // tslint:disable-line
```

Comment:
When is this happening?

Reply:
Was improved in singleton `mendix-data-source-helper` module